library/core/src/fmt/mod.rs

```diff
@@ -313,6 +312,7 @@ impl<'a> Arguments<'a> {
     #[inline]
     #[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
     pub fn new_v1(pieces: &'a [&'static str], args: &'a [ArgumentV1<'a>]) -> Arguments<'a> {
+        assert!((args.len()..=args.len() + 1).contains(&pieces.len()));
```

Comment:
Why is this assert more strict than the one in the other constructor? The other version covers all that's required for safety.

Reply:
Basically because it was easy to also assert the max length here. A longer slice would indicate a logic error since more elements would go unused. The other constructor needs to be more flexible for cases like `format_args!("x{}y{0}", z)`. Should I take the more conservative approach and just ensure safety? I figured it will be optimized out anyways.

Reply:
I just noticed the `unsafe { run(...) }` calls down below, which implies that they're relying on these slices matching up as well. It may be cleaner to just make the constructors unsafe.

Reply:
So currently `new_v1` + `write` is sound but `new_v1_formatted` + `write` is unsound. If we merely add an assert in `new_v1`, that would maintain the status quo IIUC. Perhaps just `new_v1_formatted` should be `unsafe`. But then I think we would have to add an unsafe block to `format_args!`.

Reply:
There is an attribute `#[allow_internal_unsafe]` that you can use to add an unsafe block in format_args without worrying about breaking things.

Reply:
Okay, I took a stab at making `new_v1_formatted` unsafe.

Reply:
Can you do the same to new_v1? Seems like we should keep them consistent. LGTM otherwise.

Reply:
Isn't `new_v1` actually safe though?

Reply:
It is right now in theory with that assert. The existence of the `unsafe { run(...) }` down below before this PR (even though it's not reachable from this constructor) makes me worried that other bits of code here make more aggressive assumptions about the correctness of the arguments slices. Making both constructors unsafe would ensure we're handling things correctly there.

Reply:
Sounds good. Both are now unsafe.